File: src/utils/cloc.ts
#!/usr/bin/env bun
import { readFileSync } from 'node:fs'
import { join } from 'node:path'
import { Glob } from 'bun'
import { bunnyLog } from 'bunny-log'
import { cc, ptr } from 'bun:ffi'

const LANGUAGES_JSON_PATH = './src/utils/languages.json'
const languages: Record<string, LangDef> = JSON.parse(
	readFileSync(LANGUAGES_JSON_PATH, 'utf-8')
).languages

// Compile C code with simplified dynamic language support
const {
	symbols: { add_language, analyze_file, get_language_name },
} = cc({
	source: './src/utils/cloc.c',
	symbols: {
		add_language: {
			args: ['ptr', 'ptr', 'ptr', 'ptr', 'ptr'],
			returns: 'void',
		},
		analyze_file: {
			args: ['ptr', 'ptr', 'i32', 'ptr'],
			returns: 'void',
		},
		get_language_name: {
			args: ['ptr', 'ptr', 'i32'],
			returns: 'void',
		},
	},
})

interface LangDef {
	name?: string
	extensions?: string[]
	filenames?: string[]
	line_comment?: string | string[]
	multi_line_comments?: [string, string][]
}

// Initialize C language database
let langDataInitialized = false
function initLanguageDatabase() {
	if (langDataInitialized) return

	let count = 0
	for (const [key, langDef] of Object.entries(languages)) {
		const name = langDef.name || key

		// Extensions (comma-separated)
		const extensions = (langDef.extensions || []).join(',')

		// Line comment (use first one if array)
		const lineComment = Array.isArray(langDef.line_comment)
			? langDef.line_comment[0] || ''
			: langDef.line_comment || ''

		// Block comments (use first one if multiple)
		const blockComments = langDef.multi_line_comments?.[0]
		const blockStart = blockComments?.[0] || ''
		const blockEnd = blockComments?.[1] || ''

		// Add language to C database (convert strings to C pointers)
		add_language(
			ptr(new TextEncoder().encode(`${name}\0`)),
			ptr(new TextEncoder().encode(`${extensions}\0`)),
			ptr(new TextEncoder().encode(`${lineComment}\0`)),
			ptr(new TextEncoder().encode(`${blockStart}\0`)),
			ptr(new TextEncoder().encode(`${blockEnd}\0`))
		)
		count++
	}

	langDataInitialized = true
	bunnyLog.log(
		'language',
		`🗣️ Initialized ${count} language definitions from JSON`
	)
}

// --------- FORMATTERS ---------
const fmt = (n: number) => n.toLocaleString()
const fmtBytes = (b: number): string => {
	if (b < 1024) return `${b} B`
	if (b < 1048576) return `${(b / 1024).toFixed(2)} KB`
	if (b < 1073741824) return `${(b / 1048576).toFixed(2)} MB`
	return `${(b / 1073741824).toFixed(2)} GB`
}

// --------- MAIN ANALYZER ---------
export async function analyzeCodebase(dir = './dist') {
	const start = performance.now()

	// Initialize language database from JSON
	initLanguageDatabase()

	const files = [...new Glob('**/*').scanSync(dir)]
	bunnyLog.log(
		'analysis',
		`⚡ Ultra-fast analyzing ${files.length} files with parallel C-powered counting (${Object.keys(languages).length} languages supported)`
	)

	// OPTIMIZATION 1: Lightweight filtering (only skip obvious non-code files)
	const candidateFiles = files.filter((file) => {
		// Quick check - only filter out obviously non-code files to avoid over-filtering
		const lowerFile = file.toLowerCase()
		return !(
			lowerFile.endsWith('.png') ||
			lowerFile.endsWith('.jpg') ||
			lowerFile.endsWith('.gif') ||
			lowerFile.endsWith('.pdf') ||
			lowerFile.endsWith('.zip') ||
			lowerFile.endsWith('.exe') ||
			lowerFile.includes('node_modules') ||
			lowerFile.includes('.git/')
		)
	})

	bunnyLog.log(
		'analysis',
		`📁 Quick-filtered to ${candidateFiles.length} candidate files (${files.length - candidateFiles.length} obvious non-code files skipped)`
	)

	if (candidateFiles.length === 0) {
		bunnyLog.log('warning', 'No valid files found to analyze')
		return
	}

	// OPTIMIZATION 2: Fast synchronous file reading (async overhead not worth it)
	const validFiles: { path: string; buffer: Uint8Array }[] = []

	for (const file of candidateFiles) {
		try {
			const buffer = readFileSync(join(dir, file))
			validFiles.push({ path: file, buffer })
		} catch {
			// Skip files that can't be read
		}
	}

	if (validFiles.length === 0) {
		bunnyLog.log('warning', 'No valid files could be read')
		return
	}

	// OPTIMIZATION 3: Maximum efficiency with smart caching and buffer reuse
	const langStats = new Map<string, [number, number, number, number, number]>()

	// Pre-allocate reusable buffers for maximum performance
	const pathBuffer = new Uint8Array(512)
	const langBuffer = new Uint8Array(64)
	const resultBuffer = new Int32Array(5)
	const encoder = new TextEncoder()
	const decoder = new TextDecoder()

	// Cache encoded paths to avoid repeated encoding
	const pathEncodingCache = new Map<string, Uint8Array>()

	for (const { path, buffer } of validFiles) {
		// Use cached encoded path or create new one
		let encodedPath = pathEncodingCache.get(path)
		if (!encodedPath) {
			encodedPath = encoder.encode(`${path}\0`)
			pathEncodingCache.set(path, encodedPath)
		}

		// Efficiently reuse path buffer
		pathBuffer.fill(0) // Clear previous data
		pathBuffer.set(encodedPath.slice(0, 511))

		// Get language name using buffer reuse
		get_language_name(pathBuffer, langBuffer, 64)

		// Fast language extraction
		const nameEnd = langBuffer.indexOf(0)
		const langName = decoder.decode(
			langBuffer.slice(0, nameEnd > 0 ? nameEnd : 64)
		)

		if (langName === 'Unknown') continue

		// Analyze file with efficient buffer reuse
		analyze_file(pathBuffer, buffer, buffer.length, resultBuffer)

		// Efficient result aggregation
		if (!langStats.has(langName)) {
			langStats.set(langName, [0, 0, 0, 0, 0])
		}

		const stats = langStats.get(langName)
		if (!stats) continue
		stats[0]++ // files
		stats[1] += resultBuffer[0] // lines
		stats[2] += resultBuffer[1] // code
		stats[3] += resultBuffer[2] // comments
		stats[4] += resultBuffer[4] // size
	}

	// Display results
	bunnyLog.log(
		'analysis',
		'📊 Code Analysis Results (Dynamic C-powered + JSON)'
	)

	let totalF = 0
	let totalL = 0
	let totalC = 0
	let totalM = 0
	let totalS = 0

	const table = Array.from(langStats.entries())
		.map(([lang, stats]) => {
			const [files, lines, code, comments, size] = stats
			totalF += files
			totalL += lines
			totalC += code
			totalM += comments
			totalS += size

			return {
				Language: lang,
				Files: fmt(files),
				Lines: fmt(lines),
				Code: fmt(code),
				Comments: fmt(comments),
				Size: fmtBytes(size),
			}
		})
		.sort(
			(a, b) =>
				Number.parseInt(b.Code.replace(/,/g, '')) -
				Number.parseInt(a.Code.replace(/,/g, ''))
		)

	table.push({
		Language: 'Total',
		Files: fmt(totalF),
		Lines: fmt(totalL),
		Code: fmt(totalC),
		Comments: fmt(totalM),
		Size: fmtBytes(totalS),
	})

	bunnyLog.table(table)

	const pct = totalL ? 100 / totalL : 0
	const totalB = totalL - totalC - totalM // blanks
	bunnyLog.log('summary', `Files: ${fmt(totalF)}`)
	bunnyLog.log('summary', `Lines: ${fmt(totalL)}`)
	bunnyLog.log(
		'summary',
		`Code: ${fmt(totalC)} (${(totalC * pct).toFixed(1)}%)`
	)
	bunnyLog.log(
		'summary',
		`Comments: ${fmt(totalM)} (${(totalM * pct).toFixed(1)}%)`
	)
	bunnyLog.log(
		'summary',
		`Blanks: ${fmt(totalB)} (${(totalB * pct).toFixed(1)}%)`
	)
	bunnyLog.log('summary', `Size: ${fmtBytes(totalS)}`)
	bunnyLog.log('summary', `Languages: ${langStats.size}`)
	bunnyLog.log('timing', `Time: ${(performance.now() - start).toFixed(2)}ms`)
	bunnyLog.log(
		'success',
		`⚡ Optimized analysis completed! (${Object.keys(languages).length} languages available, efficient caching + buffer reuse)`
	)
}

// Cleanup function
process.on('exit', () => {
	if (langDataInitialized) {
		// cleanup_languages()
	}
})

// ---- CLI Entrypoint ----
if (import.meta.main) {
	const dir = Bun.argv[2] || './dist'
	await analyzeCodebase(dir)
}

File: src/utils/cloc.c
// Portable C code without standard library dependencies

// Dynamic language definitions from JSON
typedef struct {
    char name[64];
    char extensions[20][16];  // Max 20 extensions, 16 chars each
    int ext_count;
    char line_comment[8];
    char block_start[8];
    char block_end[8];
} dynamic_lang_t;

// Global language database (loaded from JSON)
static dynamic_lang_t g_languages[300]; // Static allocation for 300 languages
static int g_lang_count = 0;

// String utilities
static int str_len(const char* s) {
    if (!s) return 0;
    int len = 0;
    while (s[len]) len++;
    return len;
}

static void str_copy(char* dst, const char* src, int max_len) {
    int i = 0;
    while (src && src[i] && i < max_len - 1) {
        dst[i] = src[i];
        i++;
    }
    dst[i] = '\0';
}

static int str_equals_ci(const char* a, const char* b) {
    if (!a || !b) return 0;
    while (*a && *b) {
        char ca = *a >= 'A' && *a <= 'Z' ? *a + 32 : *a;
        char cb = *b >= 'A' && *b <= 'Z' ? *b + 32 : *b;
        if (ca != cb) return 0;
        a++;
        b++;
    }
    return *a == *b;
}

// Add a single language definition
void add_language(
    const char* name,
    const char* extensions,      // Comma-separated extensions
    const char* line_comment,
    const char* block_start,
    const char* block_end
) {
    if (g_lang_count >= 300) return; // Max languages reached

    dynamic_lang_t* lang = &g_languages[g_lang_count++];

    // Copy name
    str_copy(lang->name, name, 64);

    // Parse extensions (comma-separated)
    lang->ext_count = 0;
    if (extensions) {
        const char* ext_str = extensions;
        int start = 0;
        int pos = 0;

        while (ext_str[pos] && lang->ext_count < 20) {
            if (ext_str[pos] == ',' || ext_str[pos] == '\0') {
                // Extract extension
                int len = pos - start;
                if (len > 0 && len < 15) {
                    int j = 0;
                    // Add dot if not present
                    if (ext_str[start] != '.') {
                        lang->extensions[lang->ext_count][j++] = '.';
                    }
                    // Copy extension
                    for (int k = start; k < pos && j < 15; k++, j++) {
                        lang->extensions[lang->ext_count][j] = ext_str[k];
                    }
                    lang->extensions[lang->ext_count][j] = '\0';
                    lang->ext_count++;
                }
                start = pos + 1;
            }
            if (ext_str[pos]) pos++;
            else break;
        }
    }

    // Copy comment patterns
    str_copy(lang->line_comment, line_comment, 8);
    str_copy(lang->block_start, block_start, 8);
    str_copy(lang->block_end, block_end, 8);
}

// Fast language detection using dynamic database
static const dynamic_lang_t* detect_language(const char* filepath) {
    if (g_lang_count == 0) return 0; // No languages loaded

    // Find the last dot for extension
    const char* ext = 0;
    for (const char* p = filepath    ; *p; p++) {
        if (*p == '.') ext = p;
    }
    if (!ext) return 0;

    // Check each language
    for (int i = 0; i < g_lang_count; i++) {
        const dynamic_lang_t* lang = &g_languages[i];
        for (int j = 0; j < lang->ext_count; j++) {
            if (str_equals_ci(ext, lang->extensions[j])) {
                return lang;
            }
        }
    }
    return 0;
}

// Optimized byte matching
static int bytes_match(const unsigned char* buf, int pos, const char* pattern, int pattern_len, int buf_len) {
    if (!pattern || pattern_len == 0) return 0;
    if (pos + pattern_len > buf_len) return 0;

    for (int i = 0; i < pattern_len; i++) {
        if (buf[pos + i] != (unsigned char)pattern[i]) return 0;
    }
    return 1;
}

// Process a single file buffer
static void count_file_buffer(const unsigned char* buffer, int buf_size, const dynamic_lang_t* lang, int* result) {
    result[0] = result[1] = result[2] = result[3] = result[4] = 0;
    result[4] = buf_size; // size

    if (buf_size == 0) {
        result[0] = 1; // lines
        result[3] = 1; // blanks
        return;
    }

    const char* line_comment = (lang && lang->line_comment[0]) ? lang->line_comment : 0;
    const char* block_start = (lang && lang->block_start[0]) ? lang->block_start : 0;
    const char* block_end = (lang && lang->block_end[0]) ? lang->block_end : 0;

    int lc_len = str_len(line_comment);
    int bs_len = str_len(block_start);
    int be_len = str_len(block_end);

    int lines = 1;
    int code = 0;
    int comments = 0;
    int blanks = 0;
    int i = 0;
    int in_block = 0;
    int line_start = 1;
    int is_empty = 1;

    while (i < buf_size) {
        unsigned char c = buffer[i];

        if (c == '\n') {
            if (in_block) comments++;
            else if (is_empty) blanks++;
            else code++;

            lines++;
            line_start = 1;
            is_empty = 1;
            i++;
            continue;
        }

        if (line_start && (c == ' ' || c == '\t')) {
            i++;
            continue;
        }

        if (line_start) {
            line_start = 0;

            // Check for block comment end
            if (in_block && bytes_match(buffer, i, block_end, be_len, buf_size)) {
                in_block = 0;
                i += be_len;
                is_empty = 0;
                continue;
            }

            if (!in_block) {
                // Check for line comment
                if (bytes_match(buffer, i, line_comment, lc_len, buf_size)) {
                    comments++;
                    while (i < buf_size && buffer[i] != '\n') i++;
                    continue;
                }

                // Check for block comment start
                if (bytes_match(buffer, i, block_start, bs_len, buf_size)) {
                    in_block = 1;
                    i += bs_len;

                    // Check if block comment ends on same line
                    if (be_len > 0) {
                        for (int j = i; j <= buf_size - be_len; j++) {
                            if (buffer[j] == '\n') break;
                            if (bytes_match(buffer, j, block_end, be_len, buf_size)) {
                                in_block = 0;
                                i = j + be_len;
                                break;
                            }
                        }
                    }
                    is_empty = 0;
                    continue;
                }
            }
        }

        if (c != ' ' && c != '\t') is_empty = 0;
        i++;
    }

    // Handle final line
    if (in_block) comments++;
    else if (is_empty) blanks++;
    else code++;

    result[0] = lines;
    result[1] = code;
    result[2] = comments;
    result[3] = blanks;
}

// Analyze a single file
void analyze_file(
    const char* file_path,
    const unsigned char* file_buffer,
    int file_size,
    int* result  // Output: [lines, code, comments, blanks, size]
) {
    // Detect language
    const dynamic_lang_t* lang = detect_language(file_path);

    // Count lines in this file
    count_file_buffer(file_buffer, file_size, lang, result);
}

// Get language name for a file
void get_language_name(const char* file_path, char* lang_name, int max_len) {
    const dynamic_lang_t* lang = detect_language(file_path);
    if (lang) {
        str_copy(lang_name, lang->name, max_len);
    } else {
        str_copy(lang_name, "Unknown", max_len);
    }
}

// Cleanup function
void cleanup_languages() {
    // Legacy function - not used in new implementation
}

// Legacy functions for compatibility
void count_batch(
    const unsigned char* file_data,
    const int* file_sizes,
    const int* file_lang_offsets,
    const unsigned char* lang_data,
    const int* lang_sizes,
    int num_files,
    int* results
) {
    // Legacy function - not used in new implementation
}

void count_buffer(const unsigned char* buffer, int buf_size,
                  const unsigned char* line_comment, int lc_len,
                  const unsigned char* block_start, int bs_len,
                  const unsigned char* block_end, int be_len,
                  int* result) {
    // Legacy function - not used in new implementation
}

// BATCH PROCESSING OPTIMIZATION - Process multiple files in one C call
typedef struct {
    int files;
    int lines;
    int code;
    int comments;
    int size;
} lang_stats_t;

// Optimized batch analyzer - processes all files in one call
void analyze_batch(
    const char** file_paths,
    const unsigned char** file_buffers,
    const int* file_sizes,
    int num_files,
    char* lang_names_out,     // Output: concatenated language names (64 chars each)
    int* results_out          // Output: [lines, code, comments, blanks, size] per file
) {
    for (int i = 0; i < num_files; i++) {
        const char* file_path = file_paths[i];
        const unsigned char* buffer = file_buffers[i];
        int file_size = file_sizes[i];

        // Get language
        const dynamic_lang_t* lang = detect_language(file_path);

        // Store language name (64 chars per file)
        char* lang_name_pos = lang_names_out + (i * 64);
        if (lang) {
            str_copy(lang_name_pos, lang->name, 64);
        } else {
            str_copy(lang_name_pos, "Unknown", 64);
        }

        // Analyze file and store results (5 ints per file)
        int* result_pos = results_out + (i * 5);
        count_file_buffer(buffer, file_size, lang, result_pos);
    }
}

// Aggregate results by language in C for maximum speed
void aggregate_results(
    const char* lang_names,        // Input: language names (64 chars each)
    const int* file_results,       // Input: results per file (5 ints each)
    int num_files,
    char* unique_langs_out,        // Output: unique language names (64 chars each)
    int* lang_stats_out,           // Output: aggregated stats (5 ints each)
    int* num_langs_out             // Output: number of unique languages
) {
    lang_stats_t temp_stats[100];  // Support up to 100 languages
    char temp_names[100][64];
    int lang_count = 0;

    for (int i = 0; i < num_files; i++) {
        const char* file_lang = lang_names + (i * 64);
        const int* file_result = file_results + (i * 5);

        // Skip unknown files
        if (str_equals_ci(file_lang, "Unknown")) continue;

        // Find existing language or create new one
        int lang_idx = -1;
        for (int j = 0; j < lang_count; j++) {
            if (str_equals_ci(temp_names[j], file_lang)) {
                lang_idx = j;
                break;
            }
        }

        if (lang_idx == -1) {
            // New language
            if (lang_count < 100) {
                lang_idx = lang_count++;
                str_copy(temp_names[lang_idx], file_lang, 64);
                temp_stats[lang_idx].files = 0;
                temp_stats[lang_idx].lines = 0;
                temp_stats[lang_idx].code = 0;
                temp_stats[lang_idx].comments = 0;
                temp_stats[lang_idx].size = 0;
            } else {
                continue; // Skip if too many languages
            }
        }

        // Aggregate stats
        temp_stats[lang_idx].files++;
        temp_stats[lang_idx].lines += file_result[0];
        temp_stats[lang_idx].code += file_result[1];
        temp_stats[lang_idx].comments += file_result[2];
        temp_stats[lang_idx].size += file_result[4];
    }

    // Copy results to output buffers
    *num_langs_out = lang_count;
    for (int i = 0; i < lang_count; i++) {
        str_copy(unique_langs_out + (i * 64), temp_names[i], 64);

        int* stats_out = lang_stats_out + (i * 5);
        stats_out[0] = temp_stats[i].files;
        stats_out[1] = temp_stats[i].lines;
        stats_out[2] = temp_stats[i].code;
        stats_out[3] = temp_stats[i].comments;
        stats_out[4] = temp_stats[i].size;
    }
}
